src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,345 @@
+
+        assert(start_time <= timeout);
+
+        // allow for times to not exactly match a block
+        if (fuzzed_data_provider.ConsumeBool()) start_time += interval / 2;
+        if (fuzzed_data_provider.ConsumeBool()) timeout += interval / 2;
+
+        // this may make timeout too early; if so, don't run the test
+        if (start_time > timeout) return;
+    } else {
+        if (fuzzed_data_provider.ConsumeBool()) {
+            start_time = Consensus::BIP9Deployment::ALWAYS_ACTIVE;
+            timeout = Consensus::BIP9Deployment::NO_TIMEOUT;
+            always_active_test = true;
+        } else {
+            start_time = 1199145601; // January 1, 2008
+            timeout = 1230767999;    // December 31, 2008
+            never_active_test = true;
+        }
+    }
+
+    TestConditionChecker checker(start_time, timeout, period, threshold, bit);
+
+    // Early exit if the versions don't signal sensibly for the deployment
+    if (!checker.Condition(ver_signal)) return;
+    if (checker.Condition(ver_nosignal)) return;
+    if (ver_nosignal < 0) return;
+
+    // TOP_BITS should ensure version will be positive
+    assert(ver_signal > 0);
+
+    // Now that we have chosen time and versions, setup to mine blocks
+    Blocks blocks(block_start_time, interval, ver_signal, ver_nosignal);
+
+    /* Strategy:
+     *  * we will mine a final period worth of blocks, with
+     *    randomised signalling according to a mask
+     *  * but before we mine those blocks, we will mine some
+     *    randomised number of prior periods; with either all
+     *    or no blocks in the period signalling
+     *
+     * We establish the mask first, then consume "bools" until
+     * we run out of fuzz data to work out how many prior periods
+     * there are and which ones will signal.
+     */
+
+    // establish the mask
+    const uint32_t signalling_mask = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
+
+    // mine prior periods
+    while (fuzzed_data_provider.remaining_bytes() > 0) {
+        // all blocks in these periods either do or don't signal
+        bool signal = fuzzed_data_provider.ConsumeBool();
+        for (int b = 0; b < period; ++b) {
+            blocks.mine_block(signal);
+        }
+
+        // don't risk exceeding max_blocks or times may wrap around
+        if (blocks.size() + period*2 > max_blocks) break;
+    }
+    // NOTE: fuzzed_data_provider may be fully consumed at this point and should not be used further
+
+    // now we mine the final period and check that everything looks sane
+
+    // count the number of signalling blocks
+    int blocks_sig = 0;
+
+    // get the info for the first block of the period
+    CBlockIndex* prev = blocks.tip();
+    const int exp_since = checker.GetStateSinceHeightFor(prev);
+    const ThresholdState exp_state = checker.GetStateFor(prev);
+    BIP9Stats last_stats = checker.GetStateStatisticsFor(prev);
+
+    int prev_next_height = (prev == nullptr ? 0 : prev->nHeight + 1);
```

Comment:
style-nit: Can be const. Also any downside to assert the genesis block is always DEFINED (unless the non-BIP9 compliant special cases never active/always active)

```diff
-    int prev_next_height = (prev == nullptr ? 0 : prev->nHeight + 1);
+    const int prev_next_height = (prev == nullptr ? 0 : prev->nHeight + 1);
     assert(exp_since <= prev_next_height);
 
+    if (prev == nullptr && !always_active_test && !never_active_test) {
+        assert(exp_state == ThresholdState::DEFINED);
+    }
+

Reply:
`never_active_test` also starts in DEFINED with bip9 (though it wouldn't wth bip8)

If `prev == nullptr` then `prev_net_height == 0` and therefore `exp_since <= 0` (from the above code) and at the end there's `assert(exp_since > 0 || exp_state == ThresholdState::DEFINED)` so this case should be already covered.

Reply:
Ah, thanks.